lib/date_time/date_time_core.c

```diff
@@ -101,10 +103,11 @@ static void date_time_core_schedule_retry(void)
 		return;
 	}
 
-	if (date_time_core_schedule_work(CONFIG_DATE_TIME_RETRY_INTERVAL_SECONDS) == 0) {
-		LOG_DBG("Date time update retry in: %d seconds",
-			CONFIG_DATE_TIME_RETRY_INTERVAL_SECONDS);
-	}
+	/* Scheduling new update cannot fail because we are never doing retries
+	 * if we have fresh enough time
+	 */
+	date_time_core_schedule_work(CONFIG_DATE_TIME_RETRY_INTERVAL_SECONDS);
+	LOG_DBG("Date time update retry in: %d seconds", CONFIG_DATE_TIME_RETRY_INTERVAL_SECONDS);
```

Comment:
Suppose someone uses `date_time_core_update_async` half-way through the long interval, and the first attempt fails, then, to prevent upsetting the originally scheduled update, the retry will be cancelled. 

We could split `retry_count` into `retry_count` and `retry_count_manual`, and similarly split `reschedule_blocked`, to allow manually triggered updates and their retries to completely not interfere with automatically scheduled updates (similar to how there are already two separate work items for the two update sources).

Or we can accept this as a limitation of using `date_time_core_update_async`. I'm down for either.

But as-written, retry schedule can fail

Reply:
Ok, I'll need to create a test for this to see how it behaves.

Reply:
> Suppose someone uses `date_time_core_update_async` half-way through the long interval, and the first attempt fails, then, to prevent upsetting the originally scheduled update, the retry will be cancelled.

I have reverted this change and I still cannot make retry scheduling fail. It would require that 
`atomic_clear(&reschedule_blocked);` was not called in `date_time_update_work_fn` but that always happen if we start an update whether it's a scheduled update or one triggered by `date_time_core_update_async`.

A log from unit tests with failing `date_time_core_update_async` update causing a retry to be scheduled (see `test_date_time_update_cycle_with_failing_async`):
```
[00:00:16.180,000] <dbg> date_time: date_time_core_register_handler: Registering handler 0x804bb1f
[00:00:19.190,000] <dbg> date_time: date_time_update_work_fn: Updating date time UTC...
[00:00:19.190,000] <dbg> date_time: date_time_core_current_check: Current date time too old
[00:00:19.190,000] <dbg> date_time: date_time_update_work_fn: Getting time from cellular network
[00:00:19.190,000] <dbg> date_time: date_time_modem_get: Time obtained from cellular network
[00:00:19.190,000] <dbg> date_time: date_time_core_schedule_update: New periodic date time update in: 3 seconds
[00:00:19.190,000] <dbg> date_time: date_time_core_store_tz: System time updated: 2024-01-30 20:40:52+12
[00:00:20.200,000] <dbg> date_time: date_time_core_update_async: Requesting date-time update asynchronously
[00:00:20.200,000] <dbg> date_time: date_time_update_work_fn: Updating date time UTC...
[00:00:20.200,000] <dbg> date_time: date_time_core_current_check: Current date time too old
[00:00:20.200,000] <dbg> date_time: date_time_update_work_fn: Getting time from cellular network
[00:00:20.200,000] <wrn> date_time: Did not get time from cellular network (error: -12). This may be normal as some cellular networks don't provide it or time may not be available yet.
[00:00:20.200,000] <dbg> date_time: date_time_update_work_fn: Getting time from NTP server
[00:00:20.200,000] <err> lte_lc: Could not get registration status, error: -12
[00:00:20.200,000] <wrn> date_time: Failed getting LTE network registration status, error: -14
[00:00:20.200,000] <dbg> date_time: date_time_ntp_get: Not connected to LTE, skipping NTP UTC time update
[00:00:20.200,000] <dbg> date_time: date_time_update_work_fn: Did not get time from any time source
[00:00:20.200,000] <dbg> date_time: date_time_core_schedule_retry: Date time update retry in: 1 seconds
[00:00:21.210,000] <dbg> date_time: date_time_update_work_fn: Updating date time UTC...
[00:00:21.210,000] <dbg> date_time: date_time_core_current_check: Current date time too old
[00:00:21.210,000] <dbg> date_time: date_time_update_work_fn: Getting time from cellular network
[00:00:21.210,000] <wrn> date_time: Did not get time from cellular network (error: -12). This may be normal as some cellular networks don't provide it or time may not be available yet.
[00:00:21.210,000] <dbg> date_time: date_time_update_work_fn: Getting time from NTP server
[00:00:21.210,000] <err> lte_lc: Could not get registration status, error: -12
[00:00:21.210,000] <wrn> date_time: Failed getting LTE network registration status, error: -14
[00:00:21.210,000] <dbg> date_time: date_time_ntp_get: Not connected to LTE, skipping NTP UTC time update
[00:00:21.210,000] <dbg> date_time: date_time_update_work_fn: Did not get time from any time source
[00:00:21.210,000] <dbg> date_time: date_time_core_schedule_retry: Date time update retry in: 1 seconds
[00:00:22.220,000] <dbg> date_time: date_time_update_work_fn: Updating date time UTC...
[00:00:22.220,000] <dbg> date_time: date_time_core_current_check: Current date time too old
[00:00:22.220,000] <dbg> date_time: date_time_update_work_fn: Getting time from cellular network
[00:00:22.220,000] <wrn> date_time: Did not get time from cellular network (error: -12). This may be normal as some cellular networks don't provide it or time may not be available yet.
[00:00:22.220,000] <dbg> date_time: date_time_update_work_fn: Getting time from NTP server
[00:00:22.220,000] <err> lte_lc: Could not get registration status, error: -12
[00:00:22.220,000] <wrn> date_time: Failed getting LTE network registration status, error: -14
[00:00:22.220,000] <dbg> date_time: date_time_ntp_get: Not connected to LTE, skipping NTP UTC time update
[00:00:22.220,000] <dbg> date_time: date_time_update_work_fn: Did not get time from any time source
[00:00:22.220,000] <dbg> date_time: date_time_core_schedule_update: New periodic date time update in: 3 seconds
[00:00:25.230,000] <dbg> date_time: date_time_update_work_fn: Updating date time UTC...
[00:00:25.230,000] <dbg> date_time: date_time_core_current_check: Current date time too old
[00:00:25.230,000] <dbg> date_time: date_time_update_work_fn: Getting time from cellular network
[00:00:25.230,000] <dbg> date_time: date_time_modem_get: Time obtained from cellular network
[00:00:25.230,000] <dbg> date_time: date_time_core_schedule_update: New periodic date time update in: 3 seconds
[00:00:25.230,000] <dbg> date_time: date_time_core_store_tz: System time updated: 2024-01-30 20:40:52+12
[00:00:25.230,000] <dbg> date_time: date_time_core_register_handler: Previously registered handler 0 de-registered
src/main.c:727:test_date_time_update_cycle_with_failing_async:PASS
```

Reply:
You are 100% correct, I apologize for the misdirection. My own comments in the source code even outright confirm this:

```
/* Unblock update reschedules since we are now performing an update.
* There may still be a scheduled update beyond this one (for example if this was invoked
* by date_time_core_update_async), but it will be overridden when this update triggers a
* reschedule.
*/
```

I truly don't know how I got myself so twisted around!